File: app/src/main/java/ru/evgeniy/nytimes/data/network/RestApi.java
package ru.evgeniy.nytimes.data.network;

import android.support.annotation.NonNull;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import ru.evgeniy.nytimes.BuildConfig;
import ru.evgeniy.nytimes.data.network.Interceptors.ApiKeyInterceptor;

public class RestApi {

    private static RestApi restApi;
    private final Api mApi;


    public static synchronized RestApi getInstance() {
        if (restApi == null){
            restApi = new RestApi();
        }
        return restApi;
    }

    @NonNull
    public Api getApi(){
        return mApi;
    }


    private RestApi(){
        final OkHttpClient okHttpClient = buildOkHttpClient();
        final Retrofit retrofit = buildRetrofit(okHttpClient);
        mApi = retrofit.create(Api.class);
    }

    @NonNull
    private OkHttpClient buildOkHttpClient(){
        return new OkHttpClient.Builder()
                .connectTimeout(2, TimeUnit.SECONDS)
                .addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
                .addInterceptor(new ApiKeyInterceptor())
                .build();
    }


    @NonNull
    private Retrofit buildRetrofit(@NonNull OkHttpClient client){
        return new Retrofit.Builder()
                .baseUrl(BuildConfig.SERVER_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }
}

File: app/src/main/java/ru/evgeniy/nytimes/data/network/RestApi.kt
package ru.evgeniy.nytimes.data.network

import java.util.concurrent.TimeUnit

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import ru.evgeniy.nytimes.BuildConfig
import ru.evgeniy.nytimes.data.network.interceptors.ApiKeyInterceptor

class RestApi private constructor() {
    val api: Api


    init {
        val okHttpClient = buildOkHttpClient()
        val retrofit = buildRetrofit(okHttpClient)
        api = retrofit.create(Api::class.java)
    }

    private fun buildOkHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
                .connectTimeout(2, TimeUnit.SECONDS)
                .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
                .addInterceptor(ApiKeyInterceptor())
                .build()
    }


    private fun buildRetrofit(client: OkHttpClient): Retrofit {
        return Retrofit.Builder()
                .baseUrl(BuildConfig.SERVER_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()
    }

    companion object {

        private var restApi: RestApi? = null


        val instance: RestApi
            @Synchronized get() {
                if (restApi == null) {
                    restApi = RestApi()
                }
                return restApi as RestApi
            }
    }
}
